### PR TITLE
Parallel migrations

### PR DESCRIPTION
**Pull request by mcanaves**
Added multiprocess migration option which use parallel processes to speed up the migration as mentioned on #387 . It's a transposition of django-tenants solution.

### CODE REVIEW

**Comment by mcanaves** [comment]
Accidentally closed

File: docs/use.rst [hunk 1]
```diff
@@ -89,6 +89,27 @@ The options given to ``migrate_schemas`` are also passed to every ``migrate``. H
 
 ``migrate_schemas`` raises an exception when an tenant schema is missing.
 
+
+migrate_schemas in parallel
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can run tenant migrations in parallel like this:
```
Comment (bernardopires) [inline docs/use.rst]: Suggestion for the docs:

Once the number of tenants grow, migrating all the tenants can become a bottleneck. To speed up this process, you can run tenant migrations in parallel like this:

File: dts_test_project/customers/migrations/0001_initial.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
+# -*- coding: utf-8 -*-
```
Comment (bernardopires) [inline dts_test_project/customers/migrations/0001_initial.py]: Please send this in a separate patch unless this is required for parallel migrations. Same for all migration files.

File: dts_test_project/dts_test_app/migrations/0001_initial.py [hunk 3]
```diff
@@ -1,12 +1,16 @@
 # -*- coding: utf-8 -*-
+# Generated by Django 1.10.1 on 2016-09-27 11:02
 from __future__ import unicode_literals
 
-from django.db import models, migrations
+import django.db.models.deletion
```
Comment (bernardopires) [inline dts_test_project/dts_test_app/migrations/0001_initial.py]: Is this compatible with Django 1.8 and 1.9?

File: dts_test_project/dts_test_app/models.py [hunk 4]
```diff
@@ -1,4 +1,4 @@
-from django.contrib.auth.models import User
+from django.conf.global_settings import AUTH_USER_MODEL
```
Comment (bernardopires) [inline dts_test_project/dts_test_app/models.py]: Is this compatible with Django 1.8 and 1.9?

File: tenant_schemas/__init__.py [hunk 5]
```diff
@@ -1,40 +1 @@
...
-            % get_public_schema_name())
-
-    # make sure no tenant schema is in settings.PG_EXTRA_SEARCH_PATHS
-    invalid_schemas = set(settings.PG_EXTRA_SEARCH_PATHS).intersection(
-        get_tenant_model().objects.all().values_list('schema_name', flat=True))
-    if invalid_schemas:
-        raise ImproperlyConfigured(
-            "Do not include tenant schemas (%s) on PG_EXTRA_SEARCH_PATHS."
-            % list(invalid_schemas))
+default_app_config = 'tenant_schemas.apps.TenantSchemasConfig'
```
Comment (bernardopires) [inline tenant_schemas/__init__.py]: Please move these changes to a separate PR, much appreciated!

File: tenant_schemas/test/cases.py [hunk 6]
```diff
@@ -1,28 +1,31 @@
 from django.core.management import call_command
 from django.db import connection
-from django.test import TestCase
+from django.test import TransactionTestCase
```
Comment (bernardopires) [inline tenant_schemas/test/cases.py]: Why TransacationTestCase? And please move to a separate PR if parallel migrations does not cause this test to fail.

File: tenant_schemas/tests/test_routes.py [hunk 7]
```diff
@@ -1,34 +1,55 @@
...
         super(RoutesTestCase, cls).setUpClass()
-        settings.SHARED_APPS = ('tenant_schemas', )
+        settings.SHARED_APPS = ('tenant_schemas',
+                                'customers')
         settings.TENANT_APPS = ('dts_test_app',
                                 'django.contrib.contenttypes',
-                                'django.contrib.auth', )
+                                'django.contrib.auth',)
         settings.INSTALLED_APPS = settings.SHARED_APPS + settings.TENANT_APPS
+        cls.available_apps = settings.INSTALLED_APPS
```
Comment (bernardopires) [inline tenant_schemas/tests/test_routes.py]: Variable not used anywhere else? Please move to a separate PR if parallel migrations does not cause this test to fail.

File: tenant_schemas/tests/test_tenants.py [hunk 8]
```diff
@@ -1,60 +1,109 @@
-import json
-from StringIO import StringIO
-
 from django.conf import settings
 from django.contrib.auth.models import User
-from django.core.management import call_command
-from django.db import connection
+from django.db import connection, transaction
```
Comment (bernardopires) [inline tenant_schemas/tests/test_tenants.py]: Please move to a separate PR if parallel migrations does not cause this test to fail.

File: tenant_schemas/tests/testcases.py [hunk 9]
```diff
@@ -1,41 +1,37 @@
...
                                 'django.contrib.auth', )
         settings.INSTALLED_APPS = settings.SHARED_APPS + settings.TENANT_APPS
 
-        # Django calls syncdb by default for the test database, but we want
-        # a blank public schema for this set of tests.
-        connection.set_schema_to_public()
-        cursor = connection.cursor()
-        cursor.execute('DROP SCHEMA %s CASCADE; CREATE SCHEMA %s;'
-                       % (get_public_schema_name(), get_public_schema_name(), ))
+        cls.available_apps = settings.INSTALLED_APPS
```
Comment (bernardopires) [inline tenant_schemas/tests/testcases.py]: What is this variable for? Please move to a separate PR if parallel migrations does not cause this test to fail.

File: dts_test_project/dts_test_app/migrations/0001_initial.py (hunk 3 above)
Comment (mcanaves) [inline dts_test_project/dts_test_app/migrations/0001_initial.py]: Yes, it is.

File: dts_test_project/dts_test_app/models.py (hunk 4 above)
Comment (mcanaves) [inline dts_test_project/dts_test_app/models.py]: Yes.

File: tenant_schemas/test/cases.py (hunk 6 above)
Comment (mcanaves) [inline tenant_schemas/test/cases.py]: Parallel migrations tests fails because django.test.TestCase doesn't handle connection.close properly (because of performance reasons, it applies migrations and then does a rollback). For that reason I rewrite this to django.test.TransactionTestCase (please read more here https://docs.djangoproject.com/en/1.10/topics/testing/tools/#django.test.TransactionTestCase where it explicitly states about TestCase using commit and rollback).

File: tenant_schemas/tests/test_tenants.py (hunk 8 above)
Comment (mcanaves) [inline tenant_schemas/tests/test_tenants.py]: Needed to verify that tenant is created atomically in test test_tenant_schema_is_created_atomically.

File: tenant_schemas/tests/test_routes.py (hunk 7 above)
Comment (mcanaves) [inline tenant_schemas/tests/test_routes.py]: Is a TransactionTestCase option to enable only a subset of apps for faster tests. Is used in BaseTestCase, BaseSyncTest and RoutesTestCase.

File: tenant_schemas/tests/testcases.py (hunk 9 above)
Comment (mcanaves) [inline tenant_schemas/tests/testcases.py]: Same as above.

File: dts_test_project/customers/migrations/0001_initial.py (hunk 2 above)
Comment (mcanaves) [inline dts_test_project/customers/migrations/0001_initial.py]: Is required for parallel migrations tests.

**Comment by mcanaves** [comment]
@bernardopires updated

**Comment by xgilest** [comment]
@bernardopires how is the review going? Is something else needed?
Thank you

**Comment by bernardopires** [comment]
Hey sorry, this is just a long patch and I'm away right now so that's why it's taking me longer to approve it. Will try to get it in in the next 2 weeks. Thanks!

**Comment by AlexvEck** [comment]
@bernardopires @mcanaves Hi Guys, It's been a while since this PR was updated. I think it needs just a little bit more work before it could be merged. :)

The multiprocessing of migrations would help out many!
